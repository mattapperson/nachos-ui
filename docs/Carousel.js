import React from "react";
import { View, Image } from "react-native";

import Carousel from "../src/Carousel";
import H4 from "../src/typography/H4";

// WEBPACK (CarouselExample)
const CarouselExample = () => {
  var images = [
    "https://placehold.it/300/311112",
    "https://placehold.it/300/59C480",
    "https://placehold.it/300/546C80",
    "https://placehold.it/300/D58554",
    "https://placehold.it/300/F0CD9B",
    "https://placehold.it/300/311112"
  ];
  return (
    <View>
      <H4>Example: --</H4>
      <View style={{ marginVertical: 15 }}>
        <Carousel
          autoplay
          autoplayTimeout={5000}
          loop
          index={0}
          width={400}
          height={400}
        >
          {images.map((image, index) => {
            return (
              <View key={index}>
                <Image
                  draggable={false}
                  style={{
                    width: 400,
                    height: 400
                  }}
                  source={{ uri: image }}
                />
              </View>
            );
          })}
        </Carousel>
      </View>
    </View>
  );
};
// WEBPACK

const description = `
| Prop | Type | Default | Description |
|------|------|---------|-------------|
| \`children\` | \`nodes\` | \`void\` | Include React components as slides instead of images  |
| \`style\` | \`object, array\` | \`void\` | Style of the Carousel compoment |
| \`hideIndicators\` | \`bool\` | \`false\` | Determines whether the indicator should be visible |
| \`indicatorSize\` | \`number\` | \`24\` | Size of indicator |
| \`indicatorStyle\` | \`object, array\` | \`{ color: '#bdc1cc' }\` | Style of the indicator |
| \`indicatorActiveStyle\` | \`object, array\` | \`{ color: '#2f8cff' }\` | Active style of the indicator |
| \`onChange\` | \`func\` | \`void\` | Called when a slide has been changed. It returns the index of an active slide. |
| \`width\` | \`number\` | \`300\` | Width of the Carousel |
| \`height\` | \`number\` | \`300\` | Height of the Carousel |
| \`images\` | \`array\` | \`[]\` | Array of image uri |
| \`theme\` | \`object\` | \`defaultTheme\` | An object describing a style of the component. See more in Theme section |
`;

CarouselExample.styleguide = {
  ...CarouselExample.styleguide,
  title: "Carousel",
  description
};

export default CarouselExample;
