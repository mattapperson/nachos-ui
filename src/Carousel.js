import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Image, ScrollView, View, Text, Platform } from 'react-native'
import themeManager from './themeManager'

const defaultTheme = {
  CAROUSEL_WIDTH: 300,
  CAROUSEL_HEIGHT: 300,
  CAROUSEL_INDICATOR_SIZE: 24,
  CAROUSEL_INDICATOR_COLOR: '#bdc1cc',
  CAROUSEL_INDICATOR_ACTIVE_COLOR: '#2f8cff',
}

class Carousel extends Component {
  static defaultStyle = (theme = defaultTheme) => {
    return {
      base: { flex: 1, position: 'relative' },
      image: { flex: 1, resizeMode: 'cover' },
      indicator: {
        position: 'absolute',
        left: 0,
        right: 0,
        bottom: 5,
        justifyContent: 'center',
        flexDirection: 'row',
        backgroundColor: 'transparent',
      },
      indicatorItem: {
        letterSpacing: 6,
        color: theme.CAROUSEL_INDICATOR_COLOR,
      },
      indicatorActiveItem: {
        color: theme.CAROUSEL_INDICATOR_ACTIVE_COLOR,
      },
    }
  }
  
  static propTypes = {
    hideIndicators: PropTypes.bool,
    indicatorActiveStyle: PropTypes.object,
    indicatorStyle: PropTypes.object,
    indicatorSize: PropTypes.number,
    width: PropTypes.number,
    height: PropTypes.number,
    images: PropTypes.array,
    onChange: PropTypes.func,
    children: PropTypes.node,
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    theme: PropTypes.object,
  }

  static defaultProps = {
    hideIndicators: false,
    images: [],
    onChange: () => {},
  }

  state = { activeSlide: 0 }

  _renderIndicator() {
    if (this.props.hideIndicators) {
      return
    }

    const theme = this.props.theme ||
      themeManager.getStyle('Carousel')
    const baseStyle = defaultStyle(theme)
    const indicators = []
    let indicatorStyle = {}
    let content = this.props.children
    if (!content) {
      content = this.props.images
    }

    content.forEach((child, index) => {
      indicatorStyle = {
        ...baseStyle.indicatorItem,
        ...this.props.indicatorStyle,
        fontSize: (
          this.props.indicatorSize || theme.CAROUSEL_INDICATOR_SIZE
        ),
      }
      if (this.state.activeSlide === index) {
        indicatorStyle = {
          ...indicatorStyle,
          ...baseStyle.indicatorActiveItem,
          ...this.props.indicatorActiveStyle,
        }
      }

      indicators.push(
        <Text
          style={[indicatorStyle]}
          key={index}
          onPress={this._indicatorPressed.bind(this, index)}
        >
          â€¢
        </Text>
      )
    })

    return (
      <View style={[baseStyle.indicator]}>
        {indicators}
      </View>
    )
  }

  _onAnimationEnd = (e) => {
    const theme = this.props.theme ||
      themeManager.getStyle('Carousel')
    const width = this.props.width || theme.CAROUSEL_WIDTH
    const activeSlide = e.nativeEvent.contentOffset.x / width
    this.setState({ activeSlide })

    if (this.props.onChange) {
      this.props.onChange(activeSlide)
    }
  }

  _indicatorPressed(activeSlide) {
    this.setState({ activeSlide })
    const theme = this.props.theme ||
      themeManager.getStyle('Carousel')
    const width = this.props.width || theme.CAROUSEL_WIDTH
    if (this._scrollView) {
      this._scrollView.scrollTo({
        x: activeSlide * width,
      })
    }
  }

  render() {
    const theme = this.props.theme ||
      themeManager.getStyle('Carousel')
    const width = this.props.width || theme.CAROUSEL_WIDTH
    const height = this.props.height || theme.CAROUSEL_HEIGHT

    const baseStyle = defaultStyle(theme)
    let content = this.props.children
    if (!content) {
      content = this.props.images.map((image, index) => {
        return (
          <Image
            key={index}
            style={[baseStyle.image, { width, height }]}
            source={{ uri: image }}
          />
        )
      })
    }

    return (
      <View
        style={[baseStyle.base, { width, height }, this.props.style]}
      >
        <ScrollView
          bounces={Platform.OS === 'ios' ? false : undefined}
          decelerationRate={Platform.OS === 'ios' ? 'fast' : undefined}
          automaticallyAdjustContentInsets={Platform.OS === 'ios' ? false : undefined}

          ref={(scrollView) => {
            this._scrollView = scrollView
          }}
          horizontal
          pagingEnabled
          removeClippedSubviews
          showsHorizontalScrollIndicator={false}
          scrollEventThrottle={200}
          height={this.props.height}
          onMomentumScrollEnd={this._onAnimationEnd}
        >
          {content}
        </ScrollView>
        {this._renderIndicator()}
      </View>
    )
  }
}

export default Carousel
