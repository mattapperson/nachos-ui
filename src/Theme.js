import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
