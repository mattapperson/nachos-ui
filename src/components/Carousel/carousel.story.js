import { storiesOf } from "../../storybook";
import { withInfo } from "../../storybook";
import React from "react";
import Carousel from "./index";
import { View, Image } from "react-native";

const Example = () => {
  var images = [
    "https://placehold.it/300/311112",
    "https://placehold.it/300/59C480",
    "https://placehold.it/300/546C80",
    "https://placehold.it/300/D58554",
    "https://placehold.it/300/F0CD9B",
    "https://placehold.it/300/311112"
  ];
  return (
    <View style={{ marginVertical: 15 }}>
      <Carousel
        autoplay
        autoplayTimeout={5000}
        loop
        index={0}
        width={400}
        height={400}
      >
        {images.map((image, index) => {
          return (
            <View key={index}>
              <Image
                draggable={false}
                style={{
                  width: 400,
                  height: 400
                }}
                source={{ uri: image }}
              />
            </View>
          );
        })}
      </Carousel>
    </View>
  );
};

storiesOf("Carousel", module).add("default", withInfo()(Example));

export default Example;
