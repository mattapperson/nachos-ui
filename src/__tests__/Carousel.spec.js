import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import Carousel from "../Carousel";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Carousel
        images={[
          "https://placehold.it/600/311112",
          "https://placehold.it/600/59C480"
        ]}
        height={400}
      />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("should hide indicators", () => {
  const component = renderer.create(
    <Provider>
      <Carousel
        images={[
          "https://placehold.it/600/311112",
          "https://placehold.it/600/59C480"
        ]}
        hideIndicators
      />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
