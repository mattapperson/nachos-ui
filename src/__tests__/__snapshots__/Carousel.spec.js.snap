// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should hide indicators 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  style={
    Object {
      "height": "400px",
      "width": "400px",
    }
  }
>
  <div
    className="foo-bar rn-WebkitOverflowScrolling-150rngu rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flex-13awgt0 rn-flexGrow-1m1wadx rn-flexShrink-1awmn5t rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-overflowX-11yh6sk rn-overflowY-buy8e9 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-19z077z rn-transform-emqnss rn-zIndex-1lgpqti"
    onResponderGrant={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={undefined}
    onResponderTerminationRequest={[Function]}
    onScroll={[Function]}
    onScrollShouldSetResponder={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    onWheel={[Function]}
    style={
      Object {
        "height": "400px",
        "width": "400px",
      }
    }
  >
    <div
      className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
      style={
        Object {
          "height": "401px",
          "width": "401px",
        }
      }
    />
  </div>
</div>
`;

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  style={
    Object {
      "height": "400px",
      "width": "400px",
    }
  }
>
  <div
    className="foo-bar rn-WebkitOverflowScrolling-150rngu rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flex-13awgt0 rn-flexGrow-1m1wadx rn-flexShrink-1awmn5t rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-overflowX-11yh6sk rn-overflowY-buy8e9 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-19z077z rn-transform-emqnss rn-zIndex-1lgpqti"
    onResponderGrant={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={undefined}
    onResponderTerminationRequest={[Function]}
    onScroll={[Function]}
    onScrollShouldSetResponder={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    onWheel={[Function]}
    style={
      Object {
        "height": "400px",
        "width": "400px",
      }
    }
  >
    <div
      className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
      style={
        Object {
          "height": "401px",
          "width": "401px",
        }
      }
    />
  </div>
</div>
`;
